model-optimizer/extensions/middle/MakeKaldiConstReshapable.py

```diff
@@ -0,0 +1,137 @@
+    if not mem_shape.out_port(0).disconnected():
+        for dest in mem_shape.out_port(0).get_destinations():
+            if dest.node.soft_get('op') == "Broadcast" and \
+                    dest.node.in_port(1).get_source().node.soft_get('value', []) == np.array([value], dtype=precision):
+                init_value_prev_lstm_output = dest.node
+                break
+
+    if init_value_prev_lstm_output == {}:
+        fill_value = Const(graph, {'name': 'fill_value/'+input_name,
+                                   'value': np.array([value], dtype=precision),
+                                   'shape': int64_array([1])}).create_node()
+        init_value_prev_lstm_output = Broadcast(graph, {'name': 'init_value/'+input_name,
+                                                        }).create_node()
+        init_value_prev_lstm_output.in_port(0).connect(fill_value.out_port(0))
+        init_value_prev_lstm_output.in_port(1).connect(mem_shape.out_port(0))
+
+    return init_value_prev_lstm_output
+
+
+class MakeKaldiConstReshapable(MiddleReplacementPattern):
+    """
+    Add broadcasting of constant nodes based on batch from Parameter node. This approach works only for Kaldi,
+    because it has the same batch in whole graph due to framework specific.
+    """
+    enabled = True
+    graph_condition = [lambda graph: graph.graph['fw'] == "kaldi"]
+
+    def run_after(self):
+        from extensions.middle.InsertSelect import AddSelectBeforeMemoryNodePattern
+        from extensions.middle.ReplaceMemoryOffsetWithSplice import ReplaceMemoryOffsetWithMemoryNodePattern
+        from extensions.middle.ReplaceSpliceNodePattern import ReplaceSpliceNodePattern
+        return [AddSelectBeforeMemoryNodePattern, ReplaceMemoryOffsetWithMemoryNodePattern,
+                ReplaceSpliceNodePattern]
+
+    def find_and_replace_pattern(self, graph: Graph):
+        params = graph.get_op_nodes(op="Parameter")
+        batch = params[0].shape[0]
+
+        # check that all Parameters have the same batch
+        for p in params:
+            assert(p.shape[0] == batch,
+                   "Parameter {} have batch different from the {}".format(p.soft_get('name', p.id),
+                                                                          params[0].soft_get('name', params[0].id)))
+
+        # make constants for initialization of ReadValue reshapable
+        reads = graph.get_op_nodes(op='ReadValue')
+        for read in reads:
+            if read.in_port(0).get_source().node.soft_get('op') == "Const":
+                const = read.in_port(0).get_source().node
+                const_shape = const.out_port(0).data.get_shape()
+                # extra check to be sure that we don't break shapes compatibility in graph
+                # in Kaldi models we have only 2 dimensions
+                # and batch should be set the same as we will get from Parameter
+                if len(const_shape) != 2 or const_shape[0] != batch:
+                    continue
```

Comment:
Should we raise an exception here? What will happen with a model conversion if one of these conditions is true?

Reply:
No, we shouldn't we just can't apply this transformation to nodes with another batch or another shape length.